calyx/src/analysis/dominator_map.rs

```diff
@@ -0,0 +1,489 @@
+            | ir::Control::Par(ir::Par { stmts, .. }) => {
+                for stmt in stmts {
+                    Self::build_exit_map(stmt, final_map);
+                }
+                let id = Self::get_guaranteed_attribute(c, NODE_ID);
+                final_map.insert(id, Self::get_final(c));
+            }
+        }
+    }
+
+    //Gets the "final" nodes in control c. This useful for getting
+    //what will be the predecessors of the next node in the control sequence.
+    fn get_final(c: &ir::Control) -> HashSet<u64> {
+        let mut hs = HashSet::new();
+        match c {
+            ir::Control::Empty(_) => panic!("To Do: deal w/ empty controls"),
+            ir::Control::Invoke(_)
+            | ir::Control::Enable(_)
+            | ir::Control::While(_) => {
+                hs.insert(Self::get_guaranteed_attribute(c, NODE_ID));
+            }
+            ir::Control::If(_) => {
+                hs.insert(Self::get_guaranteed_attribute(c, END_ID));
+            }
+            ir::Control::Seq(ir::Seq { stmts, .. }) => {
+                match (&stmts[..]).last() {
+                    None => panic!("error: empty Seq block. Run ___ "),
+                    Some(control) => return Self::get_final(control),
+                }
+            }
+            ir::Control::Par(ir::Par { stmts, .. }) => {
+                for stmt in stmts {
+                    let stmt_final = Self::get_final(stmt);
+                    hs = hs.union(&stmt_final).copied().collect()
+                }
+            }
+        }
+        hs
+    }
+
+    //Given an id and its predecessors pred, and a domination map d_map, updates
+    //d_map accordingly (i.e. the union of all dominators of the predecessors
+    //plus itself).
+    fn update_node(pred: &HashSet<u64>, id: u64, d_map: &mut DominatorMap) {
+        let mut union: HashSet<u64> = HashSet::new();
+        for id in pred.iter() {
+            if let Some(dominators) = d_map.map.get(id) {
+                union = union.union(dominators).copied().collect();
+            }
+        }
+        union.insert(id);
+        d_map.map.insert(id, union);
+    }
+
+    //Looks through each "node" in the "graph" and updates the dominators accordingly
+    fn update_map(
+        main_c: &ir::Control,
+        cur_id: u64,
+        pred: &HashSet<u64>,
+        d_map: &mut DominatorMap,
+    ) {
+        let c = match Self::get_control(cur_id, main_c) {
+            Some(control) => control,
+            None => return,
+        };
+        match c {
+            ir::Control::Empty(_) => {
+                unreachable!(
+                    "should not pattern match agaisnt empty in update_map()"
+                )
+            }
+            ir::Control::Invoke(_) | ir::Control::Enable(_) => {
+                Self::update_node(pred, cur_id, d_map);
+            }
+            ir::Control::Seq(ir::Seq { stmts, .. }) => {
+                //Could try to think a way of doing it w/o this first stuff
```

Comment:
I don't think that this `first` stuff is needed. What is the problem here?

Reply:
`update_map()` takes an` &HashMap` as its argument for `pred`. 
The first iteration, we want to pass `pred` as the `pred` argument when calling `update_map()` recursively. 
The subsequent iterations, we want to use `exits_map` to determine what to pass as the next `pred`. 
The only way we can use `exits_map` to determine `pred` for the first statement is if we had the id of the control statement directly preceding the `seq`.